File: ios/SmartPlanner/Services/NewsRepository.swift
import Foundation

protocol NewsRepositoryProtocol {
    func cachedArticles() -> [NewsArticle]
    func refreshArticles() async throws -> [NewsArticle]
}

final class NewsRepository: NewsRepositoryProtocol {
    private let service: NewsServiceProtocol
    private let cacheStore: NewsCacheStore

    init(
        service: NewsServiceProtocol = NewsService(),
        cacheStore: NewsCacheStore = NewsCacheStore()
    ) {
        self.service = service
        self.cacheStore = cacheStore
    }

    func cachedArticles() -> [NewsArticle] {
        cacheStore.load()
    }

    func refreshArticles() async throws -> [NewsArticle] {
        let articles = try await service.fetchTopStories()
        cacheStore.save(articles)
        return articles
    }
}

File: android/app/src/main/java/com/example/smartplanner/data/NewsRepository.kt
package com.example.smartplanner.data

import android.content.Context
import com.example.smartplanner.model.NewsArticle

class NewsRepository(context: Context) {
    private val cacheStore = NewsCacheStore(context)

    fun cachedArticles(): List<NewsArticle> = cacheStore.load()

    fun refreshArticles(): List<NewsArticle> {
        val articles = NewsService.fetchTopStories()
        cacheStore.save(articles)
        return articles
    }
}
